heartbeat/monitors/active/tcp/tcp.go

```diff
@@ -39,113 +43,219 @@ func init() {
 
 var debugf = logp.MakeDebug("tcp")
 
-type connURL struct {
-	Scheme string
-	Host   string
-	Ports  []uint16
-}
-
 func create(
 	name string,
 	cfg *common.Config,
 ) (jobs []jobs.Job, endpoints int, err error) {
-	config := DefaultConfig
-	if err := cfg.Unpack(&config); err != nil {
+	return createWithResolver(cfg, monitors.StdResolver)
+}
+
+// Custom resolver is useful for tests against hostnames locally where we don't want to depend on any
+// hostnames existing in test environments
+func createWithResolver(
+	cfg *common.Config,
+	resolver monitors.Resolver,
+) (jobs []jobs.Job, endpoints int, err error) {
+	jc, err := makeJobFactory(cfg, resolver)
+	if err != nil {
 		return nil, 0, err
 	}
 
-	tls, err := tlscommon.LoadTLSConfig(config.TLS)
+	jobs, err = jc.makeJobs()
 	if err != nil {
 		return nil, 0, err
 	}
 
-	defaultScheme := "tcp"
-	if tls != nil {
-		defaultScheme = "ssl"
+	return jobs, len(jc.endpoints), nil
+}
+
+// jobFactory is where most of the logic here lives. It provides a common context around
+// the complex logic of executing a TCP check.
+type jobFactory struct {
+	config        Config
+	tlsConfig     *tlscommon.TLSConfig
+	defaultScheme string
+	endpoints     []endpoint
+	dataCheck     dataCheck
+	resolver      monitors.Resolver
+}
+
+func makeJobFactory(commonCfg *common.Config, resolver monitors.Resolver) (jf *jobFactory, err error) {
```

Comment:
No need to name returned variables if you don't use it this way (single "return").